Sources/libhostmgr/BuildkiteScriptBuilder.swift

```diff
@@ -0,0 +1,166 @@
+                .map(convertDependencyToSource)
+                .joined(separator: "\n"),
+            environmentVariables
+                .sorted { $0.0 < $1.0 }
+                .filter { !$0.value.rawValue.isEmpty }
+                .map(convertEnvironmentVariableToExport)
+                .joined(separator: "\n"),
+            commands
+                .map(escapeCommand)
+                .joined(separator: "\n")
+        ].joined(separator: "\n")
+    }
+
+    /// Helper that takes a path like `~/.bashrc` and make it into a bash `source` command.
+    ///
+    /// Escapes spaces in paths automatically.
+    ///
+    /// Example:
+    ///
+    /// ```
+    /// # Given a path of `~/.bashrc`:
+    /// source ~/.bashrc
+    /// ```
+    func convertDependencyToSource(_ path: String) -> String {
+        "source \(path.escapingSpaces)".trimmingWhitespace
+    }
+
+    /// Helper that takes an environment variable key/value pair to an `export` statement.
+    ///
+    /// Automatically quote-wraps the value and escapes quotes as needed.
+    ///
+    /// Example:
+    ///
+    /// ```
+    /// # Given `foo:bar`
+    /// export foo="bar"
+    /// ```
+    func convertEnvironmentVariableToExport(_ pair: (String, Value)) -> String {
+        return "export \(pair.0)=\"\(pair.1.escapedRepresentation)\"".trimmingWhitespace
+    }
+
+    /// Helper that wraps command escape logic for shorthand use in a `map` statement.
+    func escapeCommand(_ command: Command) -> String {
+        command.escapedText.trimmingWhitespace
+    }
+
+    /// An object representing the `value` in an environment variable's key/value pair.
+    ///
+    /// Mostly just a way to organize escaping
+    struct Value: Equatable {
+        let rawValue: String
+
+        init(wrapping: String) {
+            self.rawValue = wrapping
+        }
+
+        /// An version of this value suitable for placement in a shell script
+        var escapedRepresentation: String {
+            rawValue
+                .escapingCodeQuotes
+                .escapingDoubleQuotes
+        }
+    }
+
+    /// An object representing one command in a shell script
+    struct Command {
+        /// The underlying command. If you want to handle escaping yourself, put the entire command here.
+        let command: String
+
+        /// The arguments for the command.
+        let arguments: [String]
+
+        init(command: String, arguments: [String] = []) {
+            self.command = command
+            self.arguments = arguments
+        }
+
+        /// The escaped command text, sutible for placement in a shell script
+        var escapedText: String {
+            "\(command) \(escapedArguments.joined(separator: " "))".trimmingWhitespace
+        }
+
+        /// A helper to print only the escaped arguments
+        var escapedArguments: [String] {
+            arguments.map {
+                if $0.rangeOfCharacter(from: .whitespaces) != nil {
+                    return "\"\($0)\""
```

Comment:
I think if the argument contains `"`, the generated shell command would be an invalid. Here is a test function for it:
```swift
    func testThatCommandQuotesArgumentsContainingQuote() throws {
        let command = BuildkiteScriptBuilder.Command(
            command: "buildkite-agent bootstrap",
            arguments: [#"/Users/my " user/.bashrc"#]
        )

        XCTAssertEqual(#"buildkite-agent "/Users/my \" user/.bashrc""#, command.escapedText)
    }
```

What do you think if we use this `String.spm_shellEscaped()` from [swift-tools-support-core](https://github.com/apple/swift-tools-support-core), rather than having our own implementation?

Reply:
Ah this is great! Not maintaining this part ourselves will save a ton of headaches. Addressed in 84ef54fddb47619238e8e997b4a03a5b295ce6c0.

Reply:
I used your test code, and added a test case for single-quotes too. The correct way to do it is....not what I expected!